WordPress/Classes/Utility/DiskCache.swift

```diff
@@ -0,0 +1,179 @@
+import CryptoKit
+
+/// An LRU disk cache that stores data in files on disk.
+///
+/// ``DiskCache`` uses an LRU cleanup policy where the least recently used items
+/// are removed first). The elements stored in the cache are automatically
+/// discarded if either *cost* or *count* limit is exceeded.
+///
+/// ``DiskCache`` is thread-safe; both reads and writes can be executed in parallel.
+///
+/// - important: It's not recommended to have more than one instance of ``DiskCache``
+/// managing the same path.
+final class DiskCache {
+    static let shared = DiskCache(
+        url: URL.getCachesURL().appendingPathComponent("WPCache", isDirectory: true)
+    )
+
+    private let rootURL: URL
+    private let sizeLimit: Int
+    private let sweepInterval: TimeInterval = 86400 // Around 1 day
+
+    /// Creates a cache instance with a given root URL.
+    ///
+    /// - parameters:
+    ///   - url: The directory URL where cached files will be stored. The
+    ///     directory will be automatically created by ``DiskCache``.
+    ///   - sizeLimit: The size limit in bytes. By default, 100 MB.
+    init(url: URL, sizeLimit: Int = 1024 * 1024 * 100) {
+        self.rootURL = url
+        self.sizeLimit = sizeLimit
+
+        do {
+            try FileManager.default.createDirectory(at: rootURL, withIntermediateDirectories: true, attributes: nil)
+        } catch {
+            DDLogError("Failed to creates cache root directory at: \(url) with error: \(error)")
+        }
+
+        performSweepIfNeeded()
+    }
+
+    /// Returns the total number of cached entities by enumerating the files.
+    func getTotalCount() throws -> Int {
+        try contents().count
+    }
+
+    // MARK: - Data
+
+    func getData(forKey key: String) -> Data? {
+        guard let url = fileURL(for: key) else { return nil }
+        return try? Data(contentsOf: url)
+    }
+
+    func setData(_ data: Data, forKey key: String) {
+        guard let url = fileURL(for: key) else { return }
+        do {
+            try data.write(to: url)
+        } catch let error as NSError {
+            guard error.code == CocoaError.fileNoSuchFile.rawValue && error.domain == CocoaError.errorDomain else { return }
+            try? FileManager.default.createDirectory(at: rootURL, withIntermediateDirectories: true, attributes: nil)
+            try? data.write(to: url) // re-create a directory and try again
+        }
+    }
+
+    func removeData(forKey key: String) {
+        guard let url = fileURL(for: key) else { return }
+        try? FileManager.default.removeItem(at: url)
+    }
+
+    func removeAll() throws {
+        try FileManager.default.removeItem(at: rootURL)
+        try FileManager.default.createDirectory(at: rootURL, withIntermediateDirectories: true, attributes: nil)
+    }
+
+    private func fileURL(for key: String) -> URL? {
+        guard !key.isEmpty else {
+            assertionFailure("The cache key can not be empty")
+            return nil
+        }
```

Comment:
IMO, we don't need to cater to a case where the function explicitly says is unacceptable. Returning a nil (which should never happen) makes every call sites need to deal with an Optional type. I'm not sure it's worth it...

Also, I think the `sha256` function can be updated to return a non-Optional type, considering sha256 checksum would never be nil and you have dealt with an exception case where the key is empty.

Reply:
The only other optional that is not handled is this: `self.data(using: .utf8)`. It's also safe to turn into an assertion because conversion to .utf8 can never fail, unlike, for example, a conversion to ascii.

I also replaced this:

```swift
SHA256.hash(data: self)
    .map { String(format: "%02x", $0) }
    .joined()
```

with this:

```swift
let digest = Insecure.SHA1.hash(data: input)
var output = ""
for byte in digest {
    output.append(String(format: "%02x", byte))
}
return output
```

I profiled it and found it to a be an order of magnitude faster.